velox/expression/ComplexProxyTypes.h

```diff
@@ -0,0 +1,218 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <iterator>
+#include <optional>
+#include "velox/common/base/Exceptions.h"
+#include "velox/core/CoreTypeSystem.h"
+#include "velox/vector/TypeAliases.h"
+#include "velox/vector/VectorTypeUtils.h"
+
+namespace facebook::velox::exec {
+
+template <typename T>
+struct VectorReader;
+
+template <typename T, typename B>
+struct VectorWriter;
+
+// Lightweight object that can be used as a proxy for array primitive elements.
+// It can writes both value and nullability, returned by operator[].
+template <typename T>
+struct PrimitiveNullableWriterProxy {
+  using vector_t = typename TypeToFlatVector<T>::type;
+  using element_t = typename CppToType<T>::NativeType;
+
+  PrimitiveNullableWriterProxy(vector_t* flatVector, vector_size_t index)
+      : flatVector_(flatVector), index_(index) {}
+  void operator=(std::nullopt_t) {
+    flatVector_->setNull(index_, true);
+  }
+
+  void operator=(element_t value) {
+    flatVector_->set(index_, value);
+  }
+
+  void operator=(const std::optional<element_t>& value) {
+    if (value) {
+      flatVector_->set(index_, value);
+    } else {
+      flatVector_->setNull(index_, true);
+    }
+  }
+
+ private:
+  vector_t* flatVector_;
+  vector_size_t index_;
+};
+
+// The object passed to the simple function interface that represent a single
+// array entry.
+// 1. General Interface:
+// - add_item()  : Add not null item and return proxy to the value to be
+// written.
+// - add_null()  : Add null item.
+// - size()     : return the size of the array.
+
+// 2. Special std::like interfaces when V is primitive:
+// - resize(n): create x items, nullability not decided yet.
+// - operator[](index): returns PrimitiveNullableWriterProxy which can be used
+// to write value or nullability at index.
+// - push_back(std::optional<v> value) : increase size by 1, adding a value
+// or null.
+template <typename V>
+class ArrayProxy {
+  using child_writer_t = VectorWriter<V, void>;
+  using element_t = typename child_writer_t::exec_out_t;
+
+ public:
+  ArrayProxy<V>(const ArrayProxy<V>&) = delete;
+
+  static bool constexpr provide_std_interface = CppToType<V>::isPrimitiveType &&
+      !std::is_same<Varchar, V>::value && !std::is_same<bool, V>::value;
+
+  // Add a new not null item to the array, increasing its size by 1.
+  FOLLY_ALWAYS_INLINE element_t& add_item() {
+    commitMostRecentChildItem();
```

Comment:
Isnt this going to be expensive, if we do this on every write ?

Reply:
For nested complex types the cost of this bool check is way cheaper than the cost of the the construction of the 
internal elements so its ok.

when the leaf element is primitive this should bo nop. since we never set 
commitLastItem, but i agree i need to make it explicit cant trust the compiler to  do so
I will change it to :
```
if constexpr (!provide_std_interface) {
    commitMostRecentChildItem();
}
```

Reply:
changes this to 
  FOLLY_ALWAYS_INLINE void commitMostRecentChildItem() {
    if constexpr (!provide_std_interface){
		if (needCommit_) {
	 		 childWriter_->commit(true);
	 		 needCommit_ = false;
			}
	}
  }